### PR TITLE
 Visium QC - Slide processing, cDNA Amplification, cDNA Analysis (x669,x671,x703)

### PR DESCRIPTION
**Pull request by seenanair**
Visium QC - Slide processing, cDNA Amplification, cDNA Analysis

### CODE REVIEW

File: src/components/slotMapper/SlotMapper.tsx [hunk 1]
```diff
@@ -263,6 +263,9 @@ function SlotMapper({
             handleCopySlots(outputAddress);
             return;
           }
+        } else {
```
Comment (jbeast) [inline src/components/slotMapper/SlotMapper.tsx:266]: Not sure I understand why this was added.

File: src/components/slotMeasurement/SlotMeasurements.tsx [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+}: SlotMeasurementProps) => {
+  const columns = React.useMemo(() => {
+    return [
+      {
+        Header: "Address",
+        id: "address",
+        accessor: (measurement: SlotMeasurementRequest) => measurement.address,
+      },
+      {
+        Header: `${measurementName}`,
```
Comment (jbeast) [inline src/components/slotMeasurement/SlotMeasurements.tsx:42]: I don't think this needs to be a template literal.

File: src/components/slotMeasurement/SlotMeasurements.tsx [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        id: measurementName,
+        Cell: ({ row }: { row: Row<SlotMeasurementRequest> }) => {
+          return (
+            <FormikInput
+              className={"rounded-md"}
+              data-testid={`measurementValue${row.index}`}
+              type={"number"}
+              label={""}
+              name={`slotMeasurements.${row.index}.value`}
+              onChange={(e: any) => {
```
Comment (jbeast) [inline src/components/slotMeasurement/SlotMeasurements.tsx:52]: Can this be something more narrow than `any`? e.g. `React.ChangeEvent`?

File: src/components/visiumQC/CDNAMeasurementQC.tsx [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+  applySameValueForAlMeasurements,
+  stepIncrement,
+  initialMeasurementVal,
+  validateMeasurementValue,
+}: {
+  measurementName: string;
+  qcType: string;
+  onSave: () => void;
+  onError: (error: ClientError) => void;
+  applySameValueForAlMeasurements: boolean;
```
Comment (jbeast) [inline src/components/visiumQC/CDNAMeasurementQC.tsx:49]: Little typo. `applySameValueForAlMeasurements` -> `applySameValueForAllMeasurements`

File: src/components/visiumQC/CDNAMeasurementQC.tsx [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+
+  /***
+   * Callback to handle Save action
+   * **/
+  const handleSave = async () => {
+    //Validate form as the Save button is not implemented as a typical Formit Submit button
+    await validateForm();
+
+    //If errors are in form data, notify user
+    if (errors.slotMeasurements && errors.slotMeasurements.length > 0) {
```
Comment (jbeast) [inline src/components/visiumQC/CDNAMeasurementQC.tsx:138]: What if there are errors on other fields that aren't slotMeasurements?

File: src/components/visiumQC/CDNAMeasurementQC.tsx [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+                  measurementName={measurementName}
+                  onChangeMeasurement={handleChangeMeasurement}
+                  validateValue={validateMeasurementValue}
+                  stepIncrement={stepIncrement ?? "1"}
+                />
+              )}
+              <div className="flex flex-col" data-testid={"labware"}>
+                <Labware
+                  labware={labwares[0]}
+                  selectable="non_empty"
```
Comment (jbeast) [inline src/components/visiumQC/CDNAMeasurementQC.tsx:273]: Does the labware need to be selectable?

File: src/pages/VisiumQC.tsx [hunk 7]
```diff
@@ -1,43 +1,54 @@
...
+  barcode: Yup.string().required().label("Barcode"),
+  slotMeasurements: Yup.array()
+    .of(
+      Yup.object().shape({
+        address: Yup.string().required(),
+        name: Yup.string().oneOf(["Cq value", "Concentration"]),
+        value: Yup.string().required(),
+      })
+    )
+    .optional(),
```
Comment (jbeast) [inline src/pages/VisiumQC.tsx:39]: Have you checked out yup's `when` method? You may be able to make this optional or required depending on which qcType is selected: https://github.com/jquense/yup#mixedwhenkeys-string--arraystring-builder-object--value-schema-schema-schema

**Review by jbeast — CHANGES_REQUESTED** [review]

File: src/components/slotMapper/SlotMapper.tsx (hunk 1 above)
Comment (seenanair) [reply src/components/slotMapper/SlotMapper.tsx:266]: Sorry, this is the fix for a bug that I happened to see while testing in Visium cDNA related to transferring samples from slide to plate (x792)

File: src/components/slotMeasurement/SlotMeasurements.tsx (hunk 2 above)
Comment (seenanair) [reply src/components/slotMeasurement/SlotMeasurements.tsx:45]: Yes, will change it.

File: src/components/visiumQC/CDNAMeasurementQC.tsx (hunk 4 above)
Comment (seenanair) [reply src/components/visiumQC/CDNAMeasurementQC.tsx]: Will do

File: src/components/visiumQC/CDNAMeasurementQC.tsx (hunk 5 above)
Comment (seenanair) [reply src/components/visiumQC/CDNAMeasurementQC.tsx]: This is handled differently in new changes.

File: src/components/visiumQC/CDNAMeasurementQC.tsx (hunk 6 above)
Comment (seenanair) [reply src/components/visiumQC/CDNAMeasurementQC.tsx:173]: No, will fix it

File: src/pages/VisiumQC.tsx (hunk 7 above)
Comment (seenanair) [reply src/pages/VisiumQC.tsx]: I will check it out.

File: src/pages/VisiumQC.tsx [hunk 8]
```diff
@@ -43,26 +43,32 @@ export interface VisiumQCFormData {
...
+        name: Yup.string().required(),
+        value: Yup.string().required(),
+      })
+    )
+    .when("qcType", {
+      is: (value: string) => value === QCType.SLIDE_PROCESSING,
+      then: Yup.array().notRequired(),
+      otherwise: Yup.array().required(),
+    }),
+});
```
Comment (jbeast) [inline src/pages/VisiumQC.tsx:68]: Ooo nice!

**Review by jbeast — APPROVED** [review]

**Pull request merged**